docs/remote-attestation.md

```diff
@@ -1,212 +1,111 @@
+T->>S: AttestationReportRequest
+S-->>S: Generate AttestationReport
+S->>T: Signed AttestationReport
+T->>U: enclave Public Key
+deactivate T
+
+Note over C,S: Fetch Enclave Public Key
+
+C->>U: GetPublicKeyRequest
+activate U
+U->>C: GetPublicKeyResponse
+deactivate U
+C-->>C: Verify Enclave Public Key attestation
+
+Note over C,S: Send Encrypted Data
+
+loop
+   C-->>C: encrypt request
+   C->>U: InvokeRequest
+   activate U
+   U->>T: InvokeRequest
+   activate T
+   T-->>T: decrypt request
+   T-->>T: process request
+   T-->>T: encrypt response
+   T->>U: InvokeResponse
+   deactivate T
+   U->>C: InvokeResponse
+   deactivate U
+   C-->>C: decrypt response
+end
+```
 
-  Trusted Runtime-->>Trusted Runtime: Generate a Signing key pair
-  Trusted Runtime-->>Trusted Runtime: Put Signing public key hash in the AttestationReport request
-  Trusted Runtime->>TEE Platform: AttestationReport request
-  TEE Platform-->>TEE Platform: Generate AttestationReport
-  TEE Platform->>Trusted Runtime: Signed AttestationReport
+### Fetch Enclave Public Key
 
-  Note over Untrusted Launcher,Trusted Runtime: Initialization
+Initially, the client connects to the untrusted launcher and requests the
+attested enclave public key. This key was generated in the enclave, but it is
+then kept in memory by the corresponding untrusted launcher, so it may be served
+to the client without having to reach into the enclave every time.
 
-  Untrusted Launcher->>Trusted Runtime: Configuration and business logic
+The client then checks:
 
-  Client-->>Client: Generate a Signing key pair
+- that the attestation report is signed by the TEE manufacturer (e.g. AMD,
+  Intel)
+- that the attestation report additional data includes the hash of the public
+  key, to confirm that the key pair was in fact generated from inside the
+  enclave
+- that the attestation report measurement corresponds to a trusted version of
+  the enclave binary
 
-  Note over Client,Trusted Runtime: Handshake Start
+If any of these checks fails, the client refuses to go ahead.
 
-  Client-->>Client: Generate an Ephemeral key pair
-  Client->>Untrusted Launcher: ClientHello
-  Untrusted Launcher->>Trusted Runtime: ClientHello
-  Trusted Runtime-->>Trusted Runtime: Generate an Ephemeral key pair
+### Send Encrypted Data
 
-  Trusted Runtime->>Untrusted Launcher: ServerIdentity
-  Untrusted Launcher->>Client: ServerIdentity
-  Client-->>Client: Validate ServerIdentity
+The client also generates an encryption key pair ahead of time (it may be reused
+across invocations, or it may be generated for each invocation).
 
-  Client->>Untrusted Launcher: ClientIdentity
-  Untrusted Launcher->>Trusted Runtime: ClientIdentity
+For each invocation (consisting of a request followed by a response) the client
+concatenates its public key with the request payload, and encrypts the resulting
+blob with the enclave public key using
+[Hybrid Encryption](https://developers.google.com/tink/hybrid).
```

Comment:
Wonder if it's worth mentioning the use of context_info as a way to bind public metadata about the request to the encrypted_request.

The framing here implies the whole request (as seen by the enclave) is private so context_info is moot

But more generally with the IDL, consider a public request message with some fields whose values are encrypted to the enclave public key. 
Other fields could be public and yet consumed by the enclave: having such fields be public helps debugging (e.g. a query of death)

Reply:
Done.